grabbit/src/main/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessor.groovy

```diff
@@ -88,10 +89,53 @@ class JcrNodesProcessor implements ItemProcessor<JcrNode, Node> {
             }
         }
 
-        final List<JcrProperty> properties = jcrNode.properties.toList()
-        Node.Builder nodeBuilder = Node.newBuilder()
+        // Hierarchy node logic
+        if(isHierarchyNode(jcrNode) || true) {
+            // It checks if the parent has required primary type as "nt:hierarchyNode"
+            // then skip it as the node has already been processed by parent node
+            if(isRequiredHierarchyChildNode(jcrNode)) {
+                return null;
+            } else {
+                // Parent "nt:hierarchyNode" type will be processed
+                return buildNode(jcrNode, getRequiredChildNodes(jcrNode))
+            }
+            // It checks if the parent has required primary type as "nt:hierarchyNode"
+            // then skip it as the node has already been processed by parent node
+            // NOTE: It is always not true for a required child node of type "nt:hierarchy" node even if parent is of type "nt:hierarchy"
+        } else if(isRequiredHierarchyChildNode(jcrNode)) {
+            return null;
+        }
+        // Non-Hierarchy node logic
+        return buildNode(jcrNode, null)
+    }
+
+    private Node buildNode(JcrNode jcrNode, Collection<JcrNode> childNodes) {
+        final Node.Builder nodeBuilder = Node.newBuilder()
         nodeBuilder.setName(jcrNode.path)
+        nodeBuilder.setProperties(buildProperties(jcrNode))
+        childNodes?.each {
+            nodeBuilder.addChildNode(buildNode(it, getRequiredChildNodes(it)))
+        }
+        return nodeBuilder.build()
+    }
 
+    private static Collection<JcrNode> getRequiredChildNodes(JcrNode jcrNode) {
+        // Here we chose not to handle the edge case of a required child hierarchy node requiring a grandchild node
+        // because no node structure exists like that within AEM.
+        return jcrNode.getNodes()
+                .findAll{JcrNode childJcrNode -> isRequiredHierarchyChildNode(childJcrNode)}
+    }
+
+    private static boolean isRequiredHierarchyChildNode(JcrNode node) {
```

Comment:
I think the method name is more appropriate as we are checking both `required` and `hierarchy`. I am not sure if there would be any condition or not but we cannot use the same method incase of non `hierarchy` `required` child node.